flow/flow.go

```diff
@@ -591,229 +608,351 @@ func SetSender(IN *Flow, portId uint8) error {
+// TODO
 func SetMerger(InArray ...*Flow) (OUT *Flow, err error) {
 	ring := low.CreateRing(generateRingName(), burstSize*sizeMultiplier)
 	for i := range InArray {
 		if err := checkFlow(InArray[i]); err != nil {
 			return nil, err
 		}
-		merge(InArray[i].current, ring)
-		InArray[i].current = nil
-		openFlowsNumber--
+		if InArray[i].line == nil {
+			merge(InArray[i].current, ring)
+			closeFlow(InArray[i])
+		} else {
+			ms := makeSlice(ring, InArray[i].line)
+			lineInsert(InArray[i], ms, true, nil)
+		}
 	}
-	OUT = new(Flow)
-	OUT.current = ring
-	openFlowsNumber++
-	return OUT, nil
+       return newFlow(ring), nil
 }
 
 // GetPortMACAddress returns default MAC address of an Ethernet port.
 func GetPortMACAddress(port uint8) [common.EtherAddrLen]uint8 {
 	return low.GetPortMACAddress(port)
 }
 
+// Service functions for Flow
+func newFlow(ring *low.Ring) *Flow {
+	OUT := new(Flow)
+	OUT.current = ring
+	openFlowsNumber++
+	return OUT
+}
+
+func newFlowLine(line *processLine, previous **Func) *Flow {
+	OUT := newFlow(nil)
+	OUT.line = line
+	OUT.previous = previous
+	return OUT
+}
+
+func finishFlow(IN *Flow) *low.Ring {
+	var ring *low.Ring
+	if IN.line == nil {
+		ring = IN.current
+               closeFlow(IN)
+	} else {
+		ring = low.CreateRing(generateRingName(), burstSize*sizeMultiplier)
+		ms := makeSlice(ring, IN.line)
+		lineInsert(IN, ms, true, nil)
+	}
+	return ring
+}
+
+func closeFlow(IN *Flow) {
+	IN.current = nil
+	IN.previous = nil
+	openFlowsNumber--
+}
+
+func lineInsert(IN *Flow, f *Func, willClose bool, context UserContext) error {
+        if err := checkFlow(IN); err != nil {
+                return err
+        }
+	if IN.line == nil {
+		IN.line = addLine(IN.current, f)
+	} else {
+		if IN.line.scalar == true && f.sFunc == nil || IN.line.scalar != true && f.sFunc != nil {
+			return common.WrapWithNFError(nil, "Scalar function to vector line", common.ScalarToVectorErr)
+		}
+		*IN.previous = f
+	}
+	if willClose {
+		closeFlow(IN)
+	} else {
+		IN.previous = &f.next[0]
+	}
+	IN.line.contexts = append(IN.line.contexts, context)
+	f.contextIndex = len(IN.line.contexts) - 1
+	return nil
+}
+
+func lineProcess(parameters interface{}, stopper chan int, report chan uint64, context []UserContext) {
+	// For scalar and vector parts
+	lp := parameters.(*lineParameters)
+	IN := lp.in
+	OUT := *lp.out
+	scalar := lp.scalar
+	outNumber := len(*lp.out)
+	InputMbufs := make([]uintptr, burstSize, burstSize)
+	OutputMbufs := make([][]uintptr, outNumber)
+	countOfPackets := make([]int, outNumber)
+	for index := range OutputMbufs {
+		OutputMbufs[index] = make([]uintptr, burstSize)
+		countOfPackets[index] = 0
+	}
+	var currentSpeed uint64
+	tick := time.Tick(time.Duration(schedTime) * time.Millisecond)
+	var pause int
+	firstFunc := lp.firstFunc
+	// For scalar part
+	var tempPacket *packet.Packet
+	var tempPacketAddr uintptr
+	// For vector part
+	tempPackets := make([]*packet.Packet, burstSize)
```

Comment:
Move those to use?

Reply:
I should create slice above loop for performance reasons